test/common/lcov.py

```diff
@@ -0,0 +1,302 @@
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+# This module can convert profile data from CDP to LCOV.
+#
+# - write_lcov (base_dir, coverage_data, outlabel)
+
+
+import json
+import os
+import sys
+import glob
+import gzip
+
+from bisect import bisect_left
+
+__all__ = (
+    "write_lcov"
+)
+
+# parse_vlq and parse_sourcemap are based on
+# https://github.com/mattrobenolt/python-sourcemap, licensed with
+# "BSD-2-Clause License"
+
+# Mapping of base64 letter -> integer value.
+B64 = dict((c, i) for i, c in
+           enumerate('ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'
+                     '0123456789+/'))
+
+
+def parse_vlq(segment):
+    """Parse a string of VLQ-encoded data.
+    Returns:
+      a list of integers.
+    """
+
+    values = []
+
+    cur, shift = 0, 0
+    for c in segment:
+        val = B64[c]
+        # Each character is 6 bits:
+        # 5 of value and the high bit is the continuation.
+        val, cont = val & 0b11111, val >> 5
+        cur += val << shift
+        shift += 5
+
+        if not cont:
+            # The low bit of the unpacked value is the sign.
+            cur, sign = cur >> 1, cur & 1
+            if sign:
+                cur = -cur
+            values.append(cur)
+            cur, shift = 0, 0
+
+    if cur or shift:
+        raise Exception('leftover cur/shift in vlq decode')
+
+    return values
+
+
+def parse_sourcemap(f, line_starts):
+    smap = json.load(f)
+    sources = smap['sources']
+    mappings = smap['mappings']
+    lines = mappings.split(';')
+
+    our_map = []
+
+    our_sources = set()
+    for s in sources:
+        if "node_modules" not in s and (s.endswith(".js") or s.endswith(".jsx")):
+            our_sources.add(s)
+
+    dst_col, src_id, src_line, src_col = 0, 0, 0, 0
+    for dst_line, line in enumerate(lines):
+        segments = line.split(',')
+        dst_col = 0
+        for segment in segments:
+            if not segment:
+                continue
+            parse = parse_vlq(segment)
+            dst_col += parse[0]
+
+            src = None
+            if len(parse) > 1:
+                src_id += parse[1]
+                src = sources[src_id]
+                src_line += parse[2]
+                src_col += parse[3]
+
+            if src in our_sources:
+                our_map.append((line_starts[dst_line] + dst_col, src, src_line))
+
+    return our_map
+
+
+class DistFile:
+    def __init__(self, path):
+        line_starts = [0]
+        for line in open(path, newline='').readlines():
+            line_starts.append(line_starts[-1] + len(line))
+        self.smap = parse_sourcemap(open(path + ".map"), line_starts)
+
+    def find_sources_slow(self, start, end):
+        res = []
+        for m in self.smap:
+            if m[0] >= start and m[0] < end:
+                res.append(m)
+        return res
```

Comment:
This can go, it's not called anywhere. Or does `find_sources()` have some disadvantage?

Reply:
It's the reference implementation, for making sure that `find_sources` works correctly.  I'll make that clear.

Reply:
Done